### ISSUE TITLE
[Bug] When using the Qwen3‑VL model for inference, the number of video_token_ids in the prepared input sequence is twice the number of actual videos provided in the inference request.

### ISSUE THREAD

**Issue by feifeialeph** [sgl-project/sglang#13805]
### Checklist

- [x] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [x] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [x] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [x] Please use English. Otherwise, it will be closed.

### Describe the bug

num of video_token_id  is 2 times of the real nums of video put in inference request, when use Qwen3-VL

### Reproduction

python3 -m sglang.launch_server \
    --model-path $MODE_PATH \
    --cuda-graph-max-bs 128 \
    --mem-fraction-static 0.8 \
    --trust-remote-code \
    --disable-radix-cache \
    --attention-backend fa3 \
    --port 30001 \
    --tp 8 \

### Environment

CUDA available: True
GPU 0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15: A100
GPU 0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15 Compute Capability: 8.0
CUDA_HOME: /usr/local/CUDA_SDK
NVCC: Cuda compilation tools, release 12.8, V12.8.
CUDA Driver Version: 1.6.1-f3016a
PyTorch: 2.7.0
sglang: 0.5.2
sgl_kernel: 0.3.9.post2
flashinfer_python: 0.2.9rc2
triton: 3.3.0+gitd63d56cc
transformers: 4.57.0
torchao: 0.11.0
numpy: 1.26.0
aiohttp: 3.10.2
fastapi: 0.116.1
hf_transfer: 0.1.9
huggingface_hub: 0.35.3
interegular: 0.3.3
modelscope: 1.29.2
orjson: 3.10.15
outlines: 0.0.37
packaging: 24.2
psutil: 5.9.8
pydantic: 2.11.7
python-multipart: 0.0.20
pyzmq: 27.1.0
uvicorn: 0.29.0
uvloop: 0.19.0
xgrammar: 0.1.21
openai: 1.90.0
tiktoken: 0.11.0
anthropic: Module Not Found
litellm: Module Not Found
decord: 0.6.0
NVIDIA Topology: 
        GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    GPU8    GPU9    GPU10   GPU11   GPU12   GPU13   GPU14    GPU15   NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    NIC6    CPU Affinity    NUMA Affinity
GPU0    X       NV2     NV1     NV2     NV1     SYS     SYS     SYS     SYS     NV1     SYS     SYS     SYS     SYS     SYS      SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     48-95,144-191           1
GPU1    NV2     X       NV2     NV1     SYS     NV1     SYS     SYS     NV1     SYS     SYS     SYS     SYS     SYS     SYS      SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     48-95,144-191           1
GPU2    NV1     NV2     X       NV1     SYS     SYS     NV2     SYS     SYS     SYS     SYS     NV1     SYS     SYS     SYS      SYS     SYS     SYS     SYS     SYS     SYS     SYS     PIX     48-95,144-191           1
GPU3    NV2     NV1     NV1     X       SYS     SYS     SYS     NV2     SYS     SYS     NV1     SYS     SYS     SYS     SYS      SYS     SYS     SYS     SYS     SYS     SYS     SYS     PIX     48-95,144-191           1
GPU4    NV1     SYS     SYS     SYS     X       NV2     NV1     NV2     SYS     SYS     SYS     SYS     SYS     NV1     SYS      SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     0-47,96-143             0
GPU5    SYS     NV1     SYS     SYS     NV2     X       NV2     NV1     SYS     SYS     SYS     SYS     NV1     SYS     SYS      SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     0-47,96-143             0
GPU6    SYS     SYS     NV2     SYS     NV1     NV2     X       NV1     SYS     SYS     SYS     SYS     SYS     SYS     SYS      NV1     SYS     SYS     SYS     SYS     PIX     SYS     SYS     0-47,96-143             0
GPU7    SYS     SYS     SYS     NV2     NV2     NV1     NV1     X       SYS     SYS     SYS     SYS     SYS     SYS     NV1      SYS     SYS     SYS     SYS     SYS     PIX     SYS     SYS     0-47,96-143             0
GPU8    SYS     NV1     SYS     SYS     SYS     SYS     SYS     SYS     X       NV2     NV1     NV2     NV1     SYS     SYS      SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     48-95,144-191           1
GPU9    NV1     SYS     SYS     SYS     SYS     SYS     SYS     SYS     NV2     X       NV2     NV1     SYS     NV1     SYS      SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     48-95,144-191           1
GPU10   SYS     SYS     SYS     NV1     SYS     SYS     SYS     SYS     NV1     NV2     X       NV1     SYS     SYS     NV2      SYS     SYS     SYS     SYS     SYS     SYS     PIX     SYS     48-95,144-191           1
GPU11   SYS     SYS     NV1     SYS     SYS     SYS     SYS     SYS     NV2     NV1     NV1     X       SYS     SYS     SYS      NV2     SYS     SYS     SYS     SYS     SYS     PIX     SYS     48-95,144-191           1
GPU12   SYS     SYS     SYS     SYS     SYS     NV1     SYS     SYS     NV1     SYS     SYS     SYS     X       NV2     NV1      NV2     SYS     SYS     SYS     SYS     SYS     SYS     SYS     0-47,96-143             0
GPU13   SYS     SYS     SYS     SYS     NV1     SYS     SYS     SYS     SYS     NV1     SYS     SYS     NV2     X       NV2      NV1     SYS     SYS     SYS     SYS     SYS     SYS     SYS     0-47,96-143             0
GPU14   SYS     SYS     SYS     SYS     SYS     SYS     SYS     NV1     SYS     SYS     NV2     SYS     NV1     NV2     XNV1     SYS     SYS     SYS     PIX     SYS     SYS     SYS     0-47,96-143             0
GPU15   SYS     SYS     SYS     SYS     SYS     SYS     NV1     SYS     SYS     SYS     SYS     NV2     NV2     NV1     NV1      X       SYS     SYS     SYS     PIX     SYS     SYS     SYS     0-47,96-143             0
NIC0    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      SYS     X       SYS     PIX     SYS     SYS     SYS     SYS     0-47,96-143             0
NIC1    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      SYS     SYS     X       SYS     SYS     SYS     SYS     SYS     48-95,144-191           1
NIC2    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      SYS     PIX     SYS     X       SYS     SYS     SYS     SYS     0-47,96-143             0
NIC3    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     PIX      PIX     SYS     SYS     SYS     X       SYS     SYS     SYS     0-47,96-143             0
NIC4    SYS     SYS     SYS     SYS     SYS     SYS     PIX     PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS      SYS     SYS     SYS     SYS     SYS     X       SYS     SYS     0-47,96-143             0
NIC5    SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     PIX     PIX     SYS     SYS     SYS      SYS     SYS     SYS     SYS     SYS     SYS     X       SYS     48-95,144-191           1
NIC6    SYS     SYS     PIX     PIX     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS     SYS      SYS     SYS     SYS     SYS     SYS     SYS     SYS     X       48-95,144-191           1

Legend:
Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_1tmp
  NIC1: mlx5_bond_0
  NIC2: mlx5_bond_1
  NIC3: mlx5_bond_2
  NIC4: mlx5_bond_3
  NIC5: mlx5_bond_4
  NIC6: mlx5_bond_5

GPU Rear Group:

  Rear ID 0: GPU 0,1,2,3,4,5,6,7
  Rear ID 1: GPU 8,9,10,11,12,13,14,15

ulimit soft: 1048576

**Comment by feifeialeph** [comment]
in  sglang/srt/layers/rotary_embedding.py   :  def get_rope_index(
        spatial_merge_size: int,
        image_token_id: int,
        video_token_id: int,
        vision_start_token_id: int,
        model_type: str,
        tokens_per_second: Optional[int] = None,
        input_ids: Optional[torch.LongTensor] = None,
        image_grid_thw: Optional[torch.LongTensor] = None,
        video_grid_thw: Optional[torch.LongTensor] = None,
        second_per_grid_ts: Optional[torch.Tensor] = None,
        **kwargs,
    ) -> Tuple[torch.Tensor, torch.Tensor]:
        if model_type == "qwen3_omni_moe":
            # For qwen3-omni
            return MRotaryEmbedding.get_rope_index_qwen3_omni(
                spatial_merge_size,
                image_token_id,
                video_token_id,
                vision_start_token_id,
                tokens_per_second,
                input_ids,
                image_grid_thw,
                video_grid_thw,
                second_per_grid_ts,
                **kwargs,
            )
        if (
            model_type.startswith("qwen3_vl") or model_type.startswith("qwen3_vl_moe")
        ) and video_grid_thw is not None:
            video_grid_thw = torch.repeat_interleave(
                video_grid_thw, video_grid_thw[:, 0], dim=0
            )
            video_grid_thw[:, 0] = 1

        mrope_position_deltas = []
        if input_ids is not None and (
            image_grid_thw is not None or video_grid_thw is not None
        ):
            total_input_ids = input_ids
            position_ids = torch.ones(
                3,
                input_ids.shape[0],
                input_ids.shape[1],
                dtype=input_ids.dtype,
                device=input_ids.device,
            )
            image_index, video_index = 0, 0
            for i, input_ids in enumerate(total_input_ids):
                image_nums, video_nums = 0, 0
                vision_start_indices = torch.argwhere(
                    input_ids == vision_start_token_id
                ).squeeze(1)
                vision_tokens = input_ids[vision_start_indices + 1]
                image_nums = (vision_tokens == image_token_id).sum()
                **video_nums = (vision_tokens == video_token_id).sum()**

**Comment by hnyls2002** [comment]
@feifeialeph Please fill out the title of this PR

**Comment by WingEdge777** [comment]
> I solved this by divide 2 for the video_num , can anybody tell me why is that? ,the video I use is not long ,just ~6s ,and[ here](https://github.com/QwenLM/Qwen3-VL?tab=readme-ov-file#online-serving) you can see the the inference request script I use [@WingEdge777](https://github.com/WingEdge777) [@JustinTong0323](https://github.com/JustinTong0323)

Hi feifeialeph.

I added some logging in the source code and had a test, but did not see anything abnormal from my point. . 

First, our multimodel processor is kind of a preprocessor of `transformers`'s `Qwen3VLVideoProcessor`, though at the time(the processor logic is not exactly aligned with transformers right now, e.g. `IMAGE_FACTOR` is 28 for qwen2_vl while `Qwen3VLVideoProcessor`'s is 32. But at first, it starts from supporting `qwen2_vl`). The `transformers`'s `Qwen3VLVideoProcessor` or we can say qwenvl's vision module decides the token nums. 

According to my test, a preprocessed video (shape: [16, 3, 756, 756 ]) plus text prompt input. `Qwen3VLVideoProcessor` output `input_ids` with length `4692`. The length is determined before the real calculation starts. The latter `get_rope_index` is just for rope calculation aligned with the video feature's length. For understanding the exact calculation logic of how video pixel number is transformed into token length, you may want to check [transformers](https://github.com/huggingface/transformers/blob/cac0a28c83cf87b7a05495de3177099c635ba852/src/transformers/models/qwen3_vl/processing_qwen3_vl.py#L195) and [here](https://github.com/huggingface/transformers/blob/cac0a28c83cf87b7a05495de3177099c635ba852/src/transformers/models/qwen3_vl/video_processing_qwen3_vl.py#L261)

Regarding my test, the video shape I used after preprocessed is `[16, 3, 756, 756]`, then became `[16,3,768,768]` after `Qwen3VLVideoProcessor`. The final total number of tokens = `16/2 * 768/16 * 768/16 / 4 + text_len = 4608.0 + text_len = 4692` which is reasonable to my understanding.

In conclusion, twice the number of actual videos provided problem does not exist if I'm not wrong.

BTW, the current preprocess (default smart size config) may not be exactly aligned for `qwen3_vl`.  Should we add some logic or do something beforehand(welcome PR)? And cc @JustinTong0323 @mickqian @yhyang201 may have more detailed elaboration for this.

Logs:
```bash
processed video shape :  torch.Size([16, 3, 756, 756])
Qwen3VLVideoProcessor resized videos shape :  torch.Size([1, 16, 3, 768, 768])
Qwen3VLVideoProcessor  pixel_values_videos shape torch.Size([18432, 1536])
Qwen3VLVideoProcessor  video_grid_thw tensor([[ 8, 48, 48]])
procesor ret input_ids  torch.Size([1, 4692])
mm_items [MultimodalDataItem(modality=<Modality.VIDEO: 3>, hash=None, pad_value=None, offsets=[(22, 597), (606, 1181), (1190, 1765), (1774, 2349), (2358, 2933), (2942, 3517), (3526, 4101), (4110, 4685)], feature=tensor([[1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
        ...,
        [1.0000, 1.0000, 1.0000,  ..., 0.4196, 0.4196, 0.4196],
        [1.0000, 1.0000, 1.0000,  ..., 0.4196, 0.4196, 0.4196],
        [1.0000, 1.0000, 1.0000,  ..., 0.4196, 0.4196, 0.4196]]), precomputed_embeddings=None, model_specific_data={'video_grid_thw': tensor([[ 8, 48, 48]])})]
before rope input_ids shape:   torch.Size([4692])
get_rope_index video_grid_thw  tensor([[ 8, 48, 48]])
get_rope_index video_grid_thw  tensor([[ 1, 48, 48],
        [ 1, 48, 48],
        [ 1, 48, 48],
        [ 1, 48, 48],
        [ 1, 48, 48],
        [ 1, 48, 48],
        [ 1, 48, 48],
        [ 1, 48, 48]])
after ropem input_ids shape :   torch.Size([4692])
mrope_positions :  torch.Size([3, 4692]) torch.Size([1, 1])
```
